src/main/java/picard/analysis/GcBiasMetricsCollector.java

```diff
@@ -102,70 +108,111 @@ public PerUnitGcBiasMetricsCollector(final String sample, final String library,
-                    type = this.sample;
-                    group = "Sample";
-                    addRead(gcData.get(type), rec, group, gc, refBases);
-                } else {
-                    type = allReads;
-                    group = "All Reads";
-                    addRead(gcData.get(type), rec, group, gc, refBases);
+                addReadToGcData(rec, this.gcData, "");
+                if (IS_NON_DUPLICATES && !rec.getDuplicateReadFlag()) {
+                    addReadToGcData(rec, this.gcDataNonDups, NON_DUPS_PREFIX);
                 }
-            }
-            else {
-                for (final Map.Entry<String, GcObject> entry : gcData.entrySet()) {
-                    final GcObject gcCur = entry.getValue();
-                    if (!rec.getReadPairedFlag() || rec.getFirstOfPairFlag()) ++gcCur.totalClusters;
+            } else {
+                updTotalClusters(rec, this.gcData);
+                if (IS_NON_DUPLICATES && !rec.getDuplicateReadFlag()) {
+                    updTotalClusters(rec, this.gcDataNonDups);
                 }
             }
         }
 
+        @Override
         public void finish() {}
 
         /////////////////////////////////////////////////////////////////////////////
+        //Called to add metrics to the output file for each level of collection
+        // these metrics are used for graphing gc bias in R script
+        /////////////////////////////////////////////////////////////////////////////
+        @Override
+        public void addMetricsToFile(final MetricsFile<GcBiasMetrics, Integer> file) {
+            addGcDataToFile(file, this.gcData, "");
+            if (IS_NON_DUPLICATES) {
+                addGcDataToFile(file, this.gcDataNonDups, NON_DUPS_PREFIX);
+            }
+        }
+
+        private void updTotalClusters(final SAMRecord rec, final Map<String, GcObject> gcData) {
+            for (final Map.Entry<String, GcObject> entry : gcData.entrySet()) {
+                final GcObject gcCur = entry.getValue();
+                if (!rec.getReadPairedFlag() || rec.getFirstOfPairFlag()) ++gcCur.totalClusters;
+            }
+        }
+
+        private Map<String, GcObject> prepareGcData() {
+            final String prefix;
+            final Map<String, GcObject> gcData = new HashMap<>();
+            if (this.readGroup != null) {
+                prefix = this.readGroup;
+                gcData.put(prefix, new GcObject());
+            } else if (this.library != null) {
+                prefix = this.library;
+                gcData.put(prefix, new GcObject());
+            } else if (this.sample != null) {
+                prefix = this.sample;
+                gcData.put(prefix, new GcObject());
+            } else {
+                prefix = allReads;
+                gcData.put(prefix, new GcObject());
+            }
+            return gcData;
+        }
+
+        private void addReadToGcData(final SAMRecord rec, final Map<String, GcObject> gcData,
```

Comment:
I think that it would be better to change the metric class and add a field that is something like 
```
Boolean INCLUDES_DUPLICATES
```
Instead of using Strings for figuring this out.